src/vs/workbench/api/common/extHostTextEditors.ts

```diff
@@ -159,4 +165,41 @@ export class ExtHostEditors implements ExtHostEditorsShape {
 	getDiffInformation(id: string): Promise<vscode.LineChange[]> {
 		return Promise.resolve(this._proxy.$getDiffInformation(id));
 	}
+
+	// --- Text editor drag and drop
+
+	async $textEditorHandleDrop(id: string, position: IPosition, dataTransferDto: DataTransferDTO, token: CancellationToken): Promise<void> {
+		const textEditor = this._extHostDocumentsAndEditors.getEditor(id);
+		if (!textEditor) {
+			throw new Error('Unknown text editor');
+		}
+
+		const pos = TypeConverters.Position.to(position);
+		const dataTransfer = DataTransferConverter.toDataTransfer(dataTransferDto);
+
+		const promises: Promise<any>[] = [];
+
+		const event = Object.freeze<vscode.TextEditorDropEvent>({
+			editor: textEditor.value,
+			position: pos,
+			dataTransfer: dataTransfer,
+			waitUntil: (p) => {
+				if (Object.isFrozen(promises)) {
+					throw illegalState('waitUntil can not be called async');
+				}
+				promises.push(Promise.resolve(p));
+			}
+		});
+
+		try {
+			this._onDidDropOnTextEditor.fire(event);
```

Comment:
fyi - there is [`AsyncEmitter`](https://github.com/microsoft/vscode/blob/d6d0bec2f28ae31dd3627a00ea2d6666dc51d726/src/vs/base/common/event.ts#L762) making some of this simpler